### ISSUE TITLE
Major issues with EC - "too few shards given" during nearly every reconstruction

### ISSUE THREAD

**Issue by Zorlin** [seaweedfs/seaweedfs#5615]
**Describe the bug**
Despite having enough shards to reconstruct my volumes, I fail to reconstruct volumes during ec.rebuild.

Example 1:
```
> ec.rebuild -force
rebuildEcVolumes collections 2
rebuildEcVolumes collection 
rebuildEcVolumes 
rebuildOneEcVolume  1572
missing shard 1572.0
vaxis.storage.riff.cc:8080 copied 1572.1 from atris.storage.riff.cc:8080
use existing shard 1572.2
vaxis.storage.riff.cc:8080 copied 1572.3 from sizer.storage.riff.cc:8080
missing shard 1572.4
missing shard 1572.5
vaxis.storage.riff.cc:8080 copied 1572.6 from al.storage.riff.cc:8080
vaxis.storage.riff.cc:8080 copied 1572.7 from sizer.storage.riff.cc:8080
vaxis.storage.riff.cc:8080 copied 1572.8 from monstar.storage.riff.cc:8080
vaxis.storage.riff.cc:8080 copied 1572.9 from ambellina.storage.riff.cc:8080
vaxis.storage.riff.cc:8080 copied 1572.10 from al.storage.riff.cc:8080
vaxis.storage.riff.cc:8080 copied 1572.11 from sizer.storage.riff.cc:8080
vaxis.storage.riff.cc:8080 copied 1572.12 from monstar.storage.riff.cc:8080
vaxis.storage.riff.cc:8080 copied 1572.13 from ambellina.storage.riff.cc:8080
delete 1572.[1 3 6 7 8 9 10 11 12 13] from vaxis.storage.riff.cc:8080
error: rpc error: code = Unknown desc = RebuildEcFiles /mnt/brick.tgc-jbod.9/seaweedfs/1572: rebuildEcFiles: reconstruct: too few shards given
```

**too few shards given** despite 3 shards being missing which shouldn't be enough to ruin that volume

A much worse example:
```
> ec.rebuild -force
rebuildEcVolumes collections 2
rebuildEcVolumes collection 
rebuildEcVolumes 
rebuildOneEcVolume  113
missing shard 113.0
vaxis.storage.riff.cc:8080 copied 113.1 from atris.storage.riff.cc:8080
vaxis.storage.riff.cc:8080 copied 113.2 from sizer.storage.riff.cc:8080
vaxis.storage.riff.cc:8080 copied 113.3 from monstar.storage.riff.cc:8080
vaxis.storage.riff.cc:8080 copied 113.4 from ambellina.storage.riff.cc:8080
use existing shard 113.5
vaxis.storage.riff.cc:8080 copied 113.6 from sizer.storage.riff.cc:8080
vaxis.storage.riff.cc:8080 copied 113.7 from monstar.storage.riff.cc:8080
vaxis.storage.riff.cc:8080 copied 113.8 from ambellina.storage.riff.cc:8080
vaxis.storage.riff.cc:8080 copied 113.9 from al.storage.riff.cc:8080
vaxis.storage.riff.cc:8080 copied 113.10 from sizer.storage.riff.cc:8080
vaxis.storage.riff.cc:8080 copied 113.11 from monstar.storage.riff.cc:8080
vaxis.storage.riff.cc:8080 copied 113.12 from ambellina.storage.riff.cc:8080
vaxis.storage.riff.cc:8080 copied 113.13 from al.storage.riff.cc:8080
delete 113.[1 2 3 4 6 7 8 9 10 11 12 13] from vaxis.storage.riff.cc:8080
error: rpc error: code = Unknown desc = RebuildEcFiles /mnt/brick.tgc-jbod.6/seaweedfs/113: rebuildEcFiles: reconstruct: too few shards given
```
In this example, we fail to reconstruct our volume even though only 1 shard is missing. In a 10+4 erasure code such as this , we should be able to lose a minimum of 4 shards at a time and not fail reconstruction.

**System Setup**
- List the command line to start "weed master", "weed volume", "weed filer", "weed s3", "weed mount".
Starting via systemd. Standard command lines as used in the docs.

- OS version
Debian 12

- output of `weed version`
`version 8000GB 3.67 d3032d1e805ab363242c833d9a61db59b3941f21 linux amd64`

- if using filer, show the content of `filer.toml`
```
root@seaweedfs-filer01:~# cat /etc/seaweedfs/filer.toml 
# Put this file to one of the location, with descending priority
#    ./filer.toml
#    $HOME/.seaweedfs/filer.toml
#    /etc/seaweedfs/filer.toml

####################################################
# Customizable filer server options
####################################################
[filer.options]
# with http DELETE, by default the filer would check whether a folder is empty.
# recursive_delete will delete all sub folders and files, similar to "rm -Rf"
recursive_delete = false
#max_file_name_length = 255

####################################################
# The following are filer store options
####################################################

[postgres]
# CREATE TABLE IF NOT EXISTS filemeta (
#   dirhash     BIGINT,
#   name        VARCHAR(65535),
#   directory   VARCHAR(65535),
#   meta        bytea,
#   PRIMARY KEY (dirhash, name)
# );
enabled = true
hostname = "redacted.riff.cc"
port = 5433
username = "seaweedfs"
password = "redacted"
database = "seaweedfs"          # create or use an existing database
schema = ""
sslmode = "disable"
connection_max_idle = 100
connection_max_open = 100
connection_max_lifetime_seconds = 0
# if insert/upsert failing, you can disable upsert or update query syntax to match your RDBMS syntax:
enableUpsert = true
#upsertQuery = """UPSERT INTO "%[1]s" (dirhash,name,directory,meta) VALUES($1,$2,$3,$4)"""

upsertQuery = """
INSERT INTO "%[1]s" (dirhash, name, directory, meta)
VALUES ($1, $2, $3, $4)
ON CONFLICT (dirhash, name)
DO UPDATE SET
    directory = EXCLUDED.directory,
    meta = EXCLUDED.meta;
"""

# Add an index:
# yugabyte=# \c seaweedfs
# You are now connected to database "seaweedfs" as user "yugabyte".
# seaweedfs=# ALTER TABLE filemeta ADD CONSTRAINT unique_dirhash UNIQUE (dirhash);
```

**Expected behavior**
EC should cleanly reconstruct.

**Additional context**
This is a major issue for us and until fixed we may have to stop using EC entirely out of safety concerns

**Comment by Zorlin** [comment]
Unless I'm misunderstanding something major, I've developed a patch that will hopefully fix this:

https://github.com/seaweedfs/seaweedfs/commit/217dfef9b7ca2bdbd95ae7cd36264132cb4038d7

I'm testing it now :) Live, on my own data!

**Comment by Zorlin** [comment]
Latest update:

Still occurring on 3.72-full-largedisk latest
```
> ec.encode -fullPercent=40 -quietFor=1s -collection=ectest -volumeId=75520
markVolumeReadonly 75520 on vaxis.storage.riff.cc:8080 ...
markVolumeReadonly 75520 on monstar.storage.riff.cc:8080 ...
markVolumeReadonly 75520 on atris.storage.riff.cc:8080 ...
generateEcShards ectest 75520 on vaxis.storage.riff.cc:8080 ...
parallelCopyEcShardsFromSource 75520 vaxis.storage.riff.cc:8080
allocate 75520.[2 7 12] vaxis.storage.riff.cc:8080 => al.storage.riff.cc:8080
allocate 75520.[4 9] vaxis.storage.riff.cc:8080 => vaxis.storage.riff.cc:8080
allocate 75520.[0 5 10] vaxis.storage.riff.cc:8080 => monstar.storage.riff.cc:8080
allocate 75520.[1 6 11] vaxis.storage.riff.cc:8080 => ambellina.storage.riff.cc:8080
allocate 75520.[3 8 13] vaxis.storage.riff.cc:8080 => atris.storage.riff.cc:8080
copy 75520.[2 7 12] vaxis.storage.riff.cc:8080 => al.storage.riff.cc:8080
mount 75520.[4 9] on vaxis.storage.riff.cc:8080
copy 75520.[0 5 10] vaxis.storage.riff.cc:8080 => monstar.storage.riff.cc:8080
copy 75520.[1 6 11] vaxis.storage.riff.cc:8080 => ambellina.storage.riff.cc:8080
copy 75520.[3 8 13] vaxis.storage.riff.cc:8080 => atris.storage.riff.cc:8080
mount 75520.[0 5 10] on monstar.storage.riff.cc:8080
I0903 16:59:31.949436 command_ec_common.go:98 vaxis.storage.riff.cc:8080 ec volume 75520 deletes shards [0 5 10]
mount 75520.[2 7 12] on al.storage.riff.cc:8080
I0903 16:59:33.488217 command_ec_common.go:98 vaxis.storage.riff.cc:8080 ec volume 75520 deletes shards [2 7 12]
mount 75520.[1 6 11] on ambellina.storage.riff.cc:8080
I0903 16:59:35.104030 command_ec_common.go:98 vaxis.storage.riff.cc:8080 ec volume 75520 deletes shards [1 6 11]
mount 75520.[3 8 13] on atris.storage.riff.cc:8080
I0903 16:59:37.708576 command_ec_common.go:98 vaxis.storage.riff.cc:8080 ec volume 75520 deletes shards [3 8 13]
unmount 75520.[0 5 10 2 7 12 1 6 11 3 8 13] from vaxis.storage.riff.cc:8080
delete 75520.[0 5 10 2 7 12 1 6 11 3 8 13] from vaxis.storage.riff.cc:8080
delete volume 75520 from vaxis.storage.riff.cc:8080
delete volume 75520 from monstar.storage.riff.cc:8080
delete volume 75520 from atris.storage.riff.cc:8080
> ec.decode
ec encode volumes: []
> ec.decode -collection=ectest -volumeId=75520
ec volume 75520 shard locations: map[al.storage.riff.cc:8080:4228 ambellina.storage.riff.cc:8080:2114 atris.storage.riff.cc:8080:8456 monstar.storage.riff.cc:8080:1057 vaxis.storage.riff.cc:8080:528]
collectEcShards: ec volume 75520 collect shards to al.storage.riff.cc:8080 from: map[al.storage.riff.cc:8080:4228 ambellina.storage.riff.cc:8080:2114 atris.storage.riff.cc:8080:8456 monstar.storage.riff.cc:8080:1057 vaxis.storage.riff.cc:8080:528]
copy 75520.[0 5] monstar.storage.riff.cc:8080 => al.storage.riff.cc:8080
mount 75520.[0 5] on al.storage.riff.cc:8080
error: collectEcShards for volume 75520: mount 75520.[0 5] on al.storage.riff.cc:8080 : rpc error: code = Unknown desc = mount 75520.0: /mnt/brick.a/seaweedfs load ec shard 75520.0: failed to create ec volume 75520: cannot open ec volume index /mnt/brick.a/seaweedfs/ectest_75520.ecx: open /mnt/brick.a/seaweedfs/ectest_75520.ecx: no such file or directory
```

Storage servers:
* al, 2 disks
* ambellina, 24 disks
* atris, 33 disks
* monstar, 24 disks
* vaxis, 33 disks

**Comment by chrislusf** [comment]
Added a fix. Please help to confirm the fix.

**Comment by Zorlin** [comment]
```
> ec.decode -collection=ectest -volumeId=75520
ec volume 75520 shard locations: map[al.storage.riff.cc:8080:4228 ambellina.storage.riff.cc:8080:2114 atris.storage.riff.cc:8080:8456 monstar.storage.riff.cc:8080:1057 vaxis.storage.riff.cc:8080:528]
collectEcShards: ec volume 75520 collect shards to atris.storage.riff.cc:8080 from: map[al.storage.riff.cc:8080:4228 ambellina.storage.riff.cc:8080:2114 atris.storage.riff.cc:8080:8456 monstar.storage.riff.cc:8080:1057 vaxis.storage.riff.cc:8080:528]
copy 75520.[4 9] vaxis.storage.riff.cc:8080 => atris.storage.riff.cc:8080
mount 75520.[4 9] on atris.storage.riff.cc:8080
copy 75520.[1 6] ambellina.storage.riff.cc:8080 => atris.storage.riff.cc:8080
mount 75520.[1 6] on atris.storage.riff.cc:8080
copy 75520.[0 5] monstar.storage.riff.cc:8080 => atris.storage.riff.cc:8080
mount 75520.[0 5] on atris.storage.riff.cc:8080
copy 75520.[2 7] al.storage.riff.cc:8080 => atris.storage.riff.cc:8080
mount 75520.[2 7] on atris.storage.riff.cc:8080
generateNormalVolume from ec volume 75520 on atris.storage.riff.cc:8080
unmount ec volume 75520 on monstar.storage.riff.cc:8080 has shards: [0 5 10]
unmount 75520.[0 5 10] from monstar.storage.riff.cc:8080
unmount ec volume 75520 on al.storage.riff.cc:8080 has shards: [2 7 12]
unmount 75520.[2 7 12] from al.storage.riff.cc:8080
unmount ec volume 75520 on atris.storage.riff.cc:8080 has shards: [0 1 2 3 4 5 6 7 8 9 13]
unmount 75520.[0 1 2 3 4 5 6 7 8 9 13] from atris.storage.riff.cc:8080
unmount ec volume 75520 on vaxis.storage.riff.cc:8080 has shards: [4 9]
unmount 75520.[4 9] from vaxis.storage.riff.cc:8080
unmount ec volume 75520 on ambellina.storage.riff.cc:8080 has shards: [1 6 11]
unmount 75520.[1 6 11] from ambellina.storage.riff.cc:8080
delete ec volume 75520 on vaxis.storage.riff.cc:8080 has shards: [4 9]
delete 75520.[4 9] from vaxis.storage.riff.cc:8080
delete ec volume 75520 on ambellina.storage.riff.cc:8080 has shards: [1 6 11]
delete 75520.[1 6 11] from ambellina.storage.riff.cc:8080
delete ec volume 75520 on monstar.storage.riff.cc:8080 has shards: [0 5 10]
delete 75520.[0 5 10] from monstar.storage.riff.cc:8080
delete ec volume 75520 on al.storage.riff.cc:8080 has shards: [2 7 12]
delete 75520.[2 7 12] from al.storage.riff.cc:8080
delete ec volume 75520 on atris.storage.riff.cc:8080 has shards: [0 1 2 3 4 5 6 7 8 9 13]
delete 75520.[0 1 2 3 4 5 6 7 8 9 13] from atris.storage.riff.cc:8080
```

😲 it works

**Comment by Zorlin** [comment]
Thank you so much for the fix!

We're now having issues getting volumes encoded:

```
> ec.encode -fullPercent=1 -collection=lokichunks
collect volumes quiet for: 3600 seconds and 1.0% full
ec encode volumes: [74756 74757 74752]
markVolumeReadonly 74756 on ambellina.storage.riff.cc:8080 ...
markVolumeReadonly 74756 on vaxis.storage.riff.cc:8080 ...
markVolumeReadonly 74756 on atris.storage.riff.cc:8080 ...
generateEcShards lokichunks 74756 on ambellina.storage.riff.cc:8080 ...
parallelCopyEcShardsFromSource 74756 ambellina.storage.riff.cc:8080
allocate 74756.[2 7 12] ambellina.storage.riff.cc:8080 => al.storage.riff.cc:8080
copy 74756.[2 7 12] ambellina.storage.riff.cc:8080 => al.storage.riff.cc:8080
allocate 74756.[4 9] ambellina.storage.riff.cc:8080 => vaxis.storage.riff.cc:8080
allocate 74756.[0 5 10] ambellina.storage.riff.cc:8080 => monstar.storage.riff.cc:8080
copy 74756.[4 9] ambellina.storage.riff.cc:8080 => vaxis.storage.riff.cc:8080
copy 74756.[0 5 10] ambellina.storage.riff.cc:8080 => monstar.storage.riff.cc:8080
allocate 74756.[3 8 13] ambellina.storage.riff.cc:8080 => atris.storage.riff.cc:8080
copy 74756.[3 8 13] ambellina.storage.riff.cc:8080 => atris.storage.riff.cc:8080
allocate 74756.[1 6 11] ambellina.storage.riff.cc:8080 => ambellina.storage.riff.cc:8080
mount 74756.[1 6 11] on ambellina.storage.riff.cc:8080
unmount 74756.[3 8 13] from atris.storage.riff.cc:8080
delete 74756.[3 8 13] from atris.storage.riff.cc:8080
unmount 74756.[4 9] from vaxis.storage.riff.cc:8080
delete 74756.[4 9] from vaxis.storage.riff.cc:8080
unmount 74756.[0 5 10] from monstar.storage.riff.cc:8080
delete 74756.[0 5 10] from monstar.storage.riff.cc:8080
unmount 74756.[1 6 11] from ambellina.storage.riff.cc:8080
delete 74756.[1 6 11] from ambellina.storage.riff.cc:8080
unmount 74756.[2 7 12] from al.storage.riff.cc:8080
delete 74756.[2 7 12] from al.storage.riff.cc:8080
error: spread ec shards for volume 74756 from ambellina.storage.riff.cc:8080: copy 74756.[4 9] ambellina.storage.riff.cc:8080 => vaxis.storage.riff.cc:8080 : rpc error: code = Unknown desc = no space left
```

but I confirmed that decoding volumes works reliably now (which is honestly the more important fix!) :)

**Comment by chrislusf** [comment]
Thanks for testing! Added another fix 8d5a6d7e74eba7f63ce6466033d24e66d19ddd32

**Comment by Zorlin** [comment]
> Thanks for testing! Added another fix [8d5a6d7](https://github.com/seaweedfs/seaweedfs/commit/8d5a6d7e74eba7f63ce6466033d24e66d19ddd32)

Yep, that seems to do it. I'm confident enough now to encode my whole FS, about 100TB, which I'll then decode back out of EC just as a final check.
